File: DP/min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost):
        # [10, 15, 20], 0 - WILL APPEND THAT O TO COST ARRAY
        cost.append(0)
        
        #start from two elements before the end, that means from 15 in above example as 20 and 0 will not get updated, because from 20 you can 
        # get to the top stair only in 1 way by selecting the 1 step which costs 20 itself, cannot do 2 step as that would be out of bound
        
        # before 20 we are going to update each element with the cost associated
        # it will be min of cost taken in 1 step and 2 step
        # lets have a look
        
        for i in range(len(cost) - 3, -1, -1):
            # cost[i] = min(cost[i] + cost[i + 1], cost[i] + cost[i + 2]), can be reduced to whats below
            cost[i] += min(cost[i + 1], cost[i + 2])
            
        return min(cost[0], cost[1])

# should return 15
print(Solution().minCostClimbingStairs( [10, 15, 20] ))

File: DP/min-cost-climbing-stairs.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int left = cost[cost.length - 1];
        int right = 0;
        
        for(int i = 1; i < cost.length; i++) {
            int newLeft = Math.min(cost[cost.length - 1 - i] + left, cost[cost.length - 1 - i] + right);
            right = left;
            left = newLeft;
        }
        
        return Math.min(left, right);
    }
}